source/common/common/radix_tree.h

```diff
@@ -0,0 +1,324 @@
+   * @param value the value to be associated with the key.
+   * @param overwrite_existing will overwrite the value when the value for a given key already
+   * exists.
+   * @return false when a value already exists for the given key.
+   */
+  bool add(absl::string_view key, Value value, bool overwrite_existing = true) {
+    // Check if the key already exists
+    Value existing = find(key);
+
+    // If a value exists and we shouldn't overwrite, return false
+    if (static_cast<bool>(existing) && !overwrite_existing) {
+      return false;
+    }
+
+    root_.insert(key, key, std::move(value));
+    return true;
+  }
+
+  /**
+   * Finds the entry associated with the key.
+   * @param key the key used to find.
+   * @param result the value associated with the key (only set if found).
+   * @return true if the key was found, false otherwise.
+   */
+  bool find(absl::string_view key, Value& result) const {
+    return root_.find_recursive(key, result);
+  }
+
+  /**
+   * Finds the entry associated with the key.
+   * @param key the key used to find.
+   * @return the Value associated with the key, or an empty-initialized Value
+   *         if there is no matching key.
+   */
+  Value find(absl::string_view key) const {
+    Value result;
+    if (find(key, result)) {
+      return result;
+    }
+    return Value{};
+  }
+
+  /**
+   * Returns the set of entries that are prefixes of the specified key, longest last.
+   * Complexity is O(min(longest key prefix, key length)).
+   * @param key the key used to find.
+   * @return a vector of values whose keys are a prefix of the specified key, longest last.
+   */
+  absl::InlinedVector<Value, 4> findMatchingPrefixes(absl::string_view key) const {
+    absl::InlinedVector<Value, 4> result;
+    absl::string_view search = key;
+    const RadixTreeNode* node = &root_;
+    bool consumed_prefix = false;
+
+    // Special case: if searching for empty string, check root node
+    if (search.empty()) {
+      if (hasValue(*node)) {
+        result.push_back(node->value_);
+      }
+      return result;
+    }
+
+    while (true) {
+      // Check if current node has a value (is a leaf) and we've consumed some prefix
+      if (hasValue(*node) && consumed_prefix) {
+        result.push_back(node->value_);
+      }
+
+      // Check for key exhaustion
+      if (search.empty()) {
+        break;
+      }
+
+      // Look for an edge
+      uint8_t firstChar = static_cast<uint8_t>(search[0]);
+      auto childIt = node->children_.find(firstChar);
+      if (childIt == node->children_.end()) {
+        break;
+      }
+
+      const RadixTreeNode& child = childIt->second;
+      node = &child;
+
+      // Consume the search prefix
+      if (search.size() >= child.prefix_.size() &&
+          search.substr(0, child.prefix_.size()) == child.prefix_) {
+        search = search.substr(child.prefix_.size());
+        consumed_prefix = true;
+      } else {
+        break;
+      }
```

Comment:
Nit: invert the condition so it's if it doesn't match, break, then you can consume the search prefix with one fewer indent.
```
  if (search.size() < child.prefix_.size() || search.substr(0, child.prefix_.size()) == child.prefix) {
    break;
  }
  // Consume the search prefix
  search = search.substr(child.prefix_.size());
```

this comment also applies to findLongestPrefix